### PR TITLE
Geocode Pickup Locations

### PR DESCRIPTION
**Pull request by seanpdoyle**
https://trello.com/c/nKmP63pg

Uses the [`geocoder`][gem] to geocode pickup locations in a background
worker.

[gem]: https://github.com/alexreisner/geocoder

### CODE REVIEW

File: app/controllers/locations_controller.rb [hunk 1]
```diff
@@ -1,25 +1,29 @@
...
+  def enqueue_geocode_job
+    GeocodeJob.perform_later(@location)
+  end
+
+  def save_location?
+    build_location.update(location_params) &&
+      @profile.update(zipcode: location_params[:zipcode])
   end
 
   def build_location
```
Comment (jsom) [inline app/controllers/locations_controller.rb]: Is it common to call a method like this `build_#{association}`? In some sense it feels more of like "I don't care if it exists or not, just give me a location associated with this profile" so I would just call the method `location`. Not important though, just my opinion.

File: app/controllers/locations_controller.rb [hunk 2]
```diff
@@ -1,25 +1,29 @@
...
     end
   end
 
   private
 
-  def update_location?
-    @location.update(location_params) &&
-      current_user.update(zipcode: location_params[:zipcode])
+  def enqueue_geocode_job
+    GeocodeJob.perform_later(@location)
```
Comment (jsom) [inline app/controllers/locations_controller.rb]: Did you want to access this directly as an instance variable? Should it use the `build_location` helper so you know `@location` has been initialised?

File: app/jobs/geocode_job.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class GeocodeJob < ActiveJob::Base
+  def perform(location)
```
Comment (jsom) [inline app/jobs/geocode_job.rb]: I haven't used ActiveJob yet -- are you able to pass instances through like that? In the past I had to pass IDs all over the place and reconstitute the object from that. This is a nice improvement if that is no longer necessary.

File: app/models/user.rb [hunk 4]
```diff
@@ -6,7 +6,6 @@ class User < ActiveRecord::Base
   has_one :location, dependent: :destroy
 
   validates :email, presence: true
-  validates :password, presence: true
```
Comment (jsom) [inline app/models/user.rb]: What's happening to password?

File: app/controllers/locations_controller.rb [hunk 5]
```diff
@@ -1,25 +1,29 @@
...
+
+    if save_location?
+      enqueue_geocode_job
 
-    if update_location?
       redirect_to profile_url
     else
-      @profile = current_user
-
       render "profiles/show"
```
Comment (jsom) [inline app/controllers/locations_controller.rb]: Was no change required in the template? Or is `profiles/show` still using the `@location` directly?

File: app/controllers/locations_controller.rb (hunk 1 above)
Comment (seanpdoyle) [inline app/controllers/locations_controller.rb]: Sounds good to me

File: app/jobs/geocode_job.rb (hunk 3 above)
Comment (seanpdoyle) [inline app/jobs/geocode_job.rb]: ActiveJob introduce (and motivated the development of) https://github.com/rails/globalid

File: app/controllers/locations_controller.rb (hunk 5 above)
Comment (seanpdoyle) [inline app/controllers/locations_controller.rb]: `profiles/show` depends on `@profile` existing (which is instantiated at the top of this method)

Comment (jsom) [inline app/controllers/locations_controller.rb]: Oh great! I see in the template you're using `@profile.location` already; I thought it was using the `@location` instance. Do you feel like this is pretty clean now?

Comment (seanpdoyle) [inline app/controllers/locations_controller.rb]: I think it's been improved greatly. Thanks!

**Comment by desi** [comment]
I looked at most of this one before. I think it LGTM

**Pull request merged**